src/azul/service/responseobjects/hca_response_v5.py

```diff
@@ -281,47 +292,139 @@ def _construct_tsv_content(self):
 
     def _construct_bdbag(self) -> str:
         """
-        Create and return a file object of the sample data.
+        Create and return a file object of the metadata coming out of ES search.
         """
         es_search = self.es_search
-        sample_file_object = StringIO()
-        sample_writer = csv.writer(sample_file_object, dialect='excel-tab')
-        # Add fieldnames.
-        sample_writer.writerow(list(chain(self.manifest_entries['contents.specimens'].keys(),
-                                          self.manifest_entries['contents.cell_suspensions'].keys(),
-                                          self.manifest_entries['bundles'].keys(),
-                                          self.manifest_entries['contents.files'].keys(),
-                                          ['file_url'], ['dos_url'])))
-        participants = set()
-        for hit in es_search.scan():
-            self._iterate_hit_bdbag(hit, participants, sample_writer)
+        bundles = BDBagOrderedDict()
+        for hit_obj in es_search.scan():
+            hit = hit_obj.to_dict()
+            bundles = self._collect_bdbag_tsv_file_headers(hit, bundles)
+        header_row = self._create_bdbag_tsv_file_header_row(bundles)
+        tsv_file_object = StringIO()
+        tsv_writer = csv.writer(tsv_file_object, dialect='excel-tab')
+        # The following change in the header is needed to satisfy Terra's entity relationship data model requirements.
+        first_col_idx = header_row.index('bundles_uuid')
+        header_row[first_col_idx] = 'entity:bundle_id'  # first column headers needs to have that format
+        header_row = self._set_as_first_col(first_col_idx, header_row)
+        col_idx = header_row.index('bundles_version')
+        header_row[col_idx] = 'bundle_version'
+        tsv_writer.writerow(header_row)
+        all_bundles = defaultdict()
+        for hit_obj in es_search.scan():
```

Comment:
And this is where I give up. In what world is it acceptable to make two passes over the ElasticSearch index, where each pass reads the entire result into memory?

Reply:
It's unacceptable - I overlooked it. Created a list of the generator to persist it in memory, and use it in the loops.